lib/new_relic/agent/method_tracer.rb

```diff
@@ -96,6 +103,260 @@ def trace_execution_unscoped(metric_names, options=NewRelic::EMPTY_HASH) #THREAD
+          end
+
+          # Checks to see if the method we are attempting to trace
+          # actually exists or not. #add_method_tracer can't do
+          # anything if the method doesn't exist.
+          def newrelic_method_exists?(method_name)
+            exists = method_defined?(method_name) || private_method_defined?(method_name)
+            ::NewRelic::Agent.logger.error("Did not trace #{derived_class_name}##{method_name} because that method does not exist") unless exists
+            exists
+          end
+
+          # Checks to see if we have already traced a method with a
+          # given metric by checking to see if the traced method
+          # exists. Warns the user if methods are being double-traced
+          # to help with debugging custom instrumentation.
+          def traced_method_exists?(method_name, metric_name_code)
+            exists = method_defined?(_traced_method_name(method_name, metric_name_code))
+            ::NewRelic::Agent.logger.error("Attempt to trace a method twice with the same metric: Method = #{method_name}, Metric Name = #{metric_name_code}") if exists
+            exists
+          end
+
+          # Returns a code snippet to be eval'd that skips tracing
+          # when the agent is not tracing execution. turns
+          # instrumentation into effectively one method call overhead
+          # when the agent is disabled
+          def assemble_code_header(method_name, metric_name_code, options)
+            header = "return #{_untraced_method_name(method_name, metric_name_code)}(*args, &block) unless NewRelic::Agent.tl_is_execution_traced?\n"
+            header += options[:code_header].to_s
+            header
+          end
+
+          # returns an eval-able string that contains the traced
+          # method code used if the agent is not creating a scope for
+          # use in scoped metrics.
+          def method_without_push_scope(method_name, metric_name_code, options)
+            define_method _traced_method_name(method_name, metric_name_code) do |*arguments|
+              eval assemble_code_header(method_name, metric_name_code, options)
+              t0 = Time.now
+              begin
+                super(*arguments)
+              ensure
+                duration = (Time.now - t0).to_f
+                NewRelic::Agent.record_metric(metric_name_code, duration)
+                eval options[:code_footer]
+              end
+            end
+          end
+
+          # returns an eval-able string that contains the tracing code
+          # for a fully traced metric including scoping
+          def method_with_push_scope(method_name, metric_name_code, options)
+            define_method _traced_method_name(method_name, metric_name_code) do |*arguments|
+              eval options[:code_header]
+              result = ::NewRelic::Agent::MethodTracerHelpers.trace_execution_scoped(metric_name_code, :metric => options[:metric]) do
+                super(*arguments)
```

Comment:
Note how we're completely sides-stepping the arguments issue with using `define_method` and calling `super` with original arguments.  We no longer have to concern ourselves with Ruby <= 2.6 vs >= 2.7